### PR TITLE
Shim gss api on Linux to delay loading libgssapi_krb5.so

### LINKED ISSUE #45720: Publishing release as single file does not include all libraries (libgssapi_krb5)
---
**Issue moved from microsoft/dotnet#1273**
- Please respond to @lastcoolnameleft.

---

_From @lastcoolnameleft on Monday, November 23, 2020 4:42:46 PM_

I'm trying to create a single-file binary to run on Linux; however, it does not include all libraries.  I'm unclear why the Kerberos library is needed.

To reproduce:
```
➜  dotnet --version
5.0.100
➜  dotnet new console -o single-file
The template "Console Application" was created successfully.

Processing post-creation actions...
Running 'dotnet restore' on single-file/single-file.csproj...
  Determining projects to restore...
  Restored /Users/thfalgou/tmp/dotnet/single-file/single-file.csproj (in 64 ms).
Restore succeeded.
➜  cd single-file 
➜  dotnet publish -r linux-x64 -p:PublishSingleFile=true -c Release
Microsoft (R) Build Engine version 16.8.0+126527ff1 for .NET
Copyright (C) Microsoft Corporation. All rights reserved.

  Determining projects to restore...
  Restored /Users/thfalgou/tmp/dotnet/single-file/single-file.csproj (in 89 ms).
  single-file -> /Users/thfalgou/tmp/dotnet/single-file/bin/Release/net5.0/linux-x64/single-file.dll
  single-file -> /Users/thfalgou/tmp/dotnet/single-file/bin/Release/net5.0/linux-x64/publish/                                                                           
➜  file bin/Release/net5.0/linux-x64/publish/single-file
bin/Release/net5.0/linux-x64/publish/single-file: ELF 64-bit LSB pie executable, x86-64, version 1 (GNU/Linux), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, for GNU/Linux 2.6.32, BuildID[sha1]=6ca69f7bc7061d8eab20013d6820ef5a5c2f7ff4, stripped, too many notes (256)
```

Start an Ubuntu instance via docker

```
➜  docker run -it -v $PWD/bin/Release/net5.0/linux-x64/publish:/tmp/single-file ubuntu bash
root@e91793747e67:/# cd /tmp/single-file/
root@e91793747e67:/tmp/single-file# ls
single-file  single-file.pdb
root@e91793747e67:/tmp/single-file# ./single-file
./single-file: error while loading shared libraries: libgssapi_krb5.so.2: cannot open shared object file: No such file or directory
root@e91793747e67:/tmp/single-file# ldd ./single-file
	linux-vdso.so.1 (0x00007ffc371c4000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f1b744cd000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f1b744c7000)
	libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007f1b744ab000)
	libgssapi_krb5.so.2 => not found
	librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007f1b744a0000)
	libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f1b742bf000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f1b7416e000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f1b74153000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f1b73f61000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f1b75126000)
```
As you can see, libgssapi_krb5.so.2 was not included.   

Due to `-p:PublishSingleFile=true`, I would expect the application to be a statically linked library instead of a dynamically linked one.

### PR DESCRIPTION
**Pull request by VSadov**
Introduced a shim for GSS APIs similar to ssl shim. There is no longer a static dependency  on `libgssapi_krb5.so` and loader does require it as a prerequisite to run singlefile apps. 

On the first use of anything from `libgssapi_krb5.so`, we load the library dynamically and set up pointers to all the functions of interest and call the functions via these indirections.

This case is simpler than ssl though - there are much less functions to shim and no gnarly issues with versioning, functions that may optional be present and so on.

Fixes: #45720 , #54347

### CODE REVIEW

**Comment by VSadov** [comment]
Here is the list of needed symbols for singlefilehost after this change:

```
readelf -d singlefilehost | grep NEEDED
 0x0000000000000001 (NEEDED)             Shared library: [libpthread.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libdl.so.2]
 0x0000000000000001 (NEEDED)             Shared library: [libz.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [librt.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libgcc_s.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libstdc++.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libm.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libc.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [ld-linux-x86-64.so.2]
```

`libgssapi_krb5.so` is no longer in the list.  
The rest seems reasonable for a minimally required set.

File: src/libraries/Native/Unix/System.Net.Security.Native/extra_libs.cmake [hunk 1]
```diff
@@ -19,5 +19,11 @@ macro(append_extra_security_libs NativeLibsExtra)
      endif()
   endif()
 
-  list(APPEND ${NativeLibsExtra} ${LIBGSS})
+  if(CLR_CMAKE_TARGET_LINUX)
+    # On Linux libgssapi_krb5.so is loaded on demand to tolerate its absense in singlefile apps that do not use it
+    list(APPEND ${NativeLibsExtra} dl)
+    add_definitions(-DGSS_DYNAMIC_LIB="libgssapi_krb5.so")
```
Comment (janvorli) [inline src/libraries/Native/Unix/System.Net.Security.Native/extra_libs.cmake:25]: I would prefer defining the name in the actual shim. We do it that way for openssl and I don't see a benefit of having it here.

File: src/libraries/Native/Unix/System.Net.Security.Native/pal_gssapi.c [hunk 2]
```diff
@@ -48,6 +53,152 @@ static gss_OID_desc gss_mech_ntlm_OID_desc = {.length = ARRAY_SIZE(gss_ntlm_oid_
...
+
+static void* volatile s_gssLib = NULL;
+
+// reference to the shim storage.
+// NOTE: the shim reference is published after all indirection pointers are initialized.
+//       when we read the indirection pointers, we do that via the shim reference.
+//       data dependency ensures that method pointers are loaded after reading and null-checking the shim reference.
+static gss_shim_t* volatile s_gss_shim_ptr = NULL;
+
+static void init_gss_shim()
```
Comment (janvorli) [inline src/libraries/Native/Unix/System.Net.Security.Native/pal_gssapi.c:109]: We could place this into a shared library constructor and then we would not have to do anything atomic or check for initialization when accessing the functions. The constructor is any void function with void argument list marked by `__attribute__((constructor))`. It can be static and may also need `__attribute__((__unused__))` so that linker doesn't eliminate it.

Comment (VSadov) [reply src/libraries/Native/Unix/System.Net.Security.Native/pal_gssapi.c:109]: Would initializing in a module constructor make it eager-initializing again?

Comment (VSadov) [reply src/libraries/Native/Unix/System.Net.Security.Native/pal_gssapi.c:109]: The goals here are: 
- tolerate cases when the krb5 .so is not present, as it happens fairly often in containers.
- delay initialization until used as this API is relatively rare on Linux.

Comment (janvorli) [reply src/libraries/Native/Unix/System.Net.Security.Native/pal_gssapi.c:109]: I am sorry, I don't know what I was thinking. Of course you are right. I wonder though - both the openssl and ICU shims have initialization functions that are explicitly called by the managed code (`CryptoNative_EnsureOpenSslInitialized` / `GlobalizationNative_LoadICU`), . The benefit is that the failure to load the library happens at a single point and not at a random call to a function from the library. It seems it would be nice to use the same way here.

Comment (VSadov) [reply src/libraries/Native/Unix/System.Net.Security.Native/pal_gssapi.c:109]: GSS/KRB5 use patterns do not seem to have explicit initialization. When used, the API is expected to be present, otherwise the app fails.

It seem acceptable and I do not think we have a too strong case to change that.

The problem with singlefile is that the app fails even when the API is _not_ used, which is a regression from non-singlefile case.
On-demand loading fixes just that part.

It is also not a mainline scenario. If the app really needs the API, the user should not have it removed.

Basically - I think a smaller change that delays the failure would be sufficient in this case.

Comment (janvorli) [reply src/libraries/Native/Unix/System.Net.Security.Native/pal_gssapi.c:109]: OpenSSL is the same case. The initialization is triggered by this code:
https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/libraries/Common/src/Interop/Android/System.Security.Cryptography.Native.Android/Interop.Initialization.cs#L30-L54
So when the System.Security.Cryptography managed assembly is loaded, the native shim is initialized by the static constructor. When the app doesn't use that assembly, the native shim is not initialized and no functions exported by the native library are called.
A benefit of this approach is that instead of abort from the native code when the library is not installed and the app wants to use it, you'll get an unhandled exception with managed stack trace.

Comment (VSadov) [reply src/libraries/Native/Unix/System.Net.Security.Native/pal_gssapi.c:109]: Ah, yes, the static constructor trick. That should be equally non-disruptive to the overall use of the API and have a better failure mode.

Let me see if we can use it here.

Comment (VSadov) [reply src/libraries/Native/Unix/System.Net.Security.Native/pal_gssapi.c]: Pushed a change that switches to static constructor scheme. 

It looks like while highly improbable, we can have concurrent initialization because `IsNtlmInstalled` could live in its own class via a file include. I can't rule out the need for atomic things, so I kept them.

File: src/libraries/Native/Unix/System.Net.Security.Native/pal_gssapi.c [hunk 3]
```diff
@@ -48,6 +53,120 @@ static gss_OID_desc gss_mech_ntlm_OID_desc = {.length = ARRAY_SIZE(gss_ntlm_oid_
...
+#undef PER_FUNCTION_BLOCK
+} gss_shim_t;
+
+static void* volatile s_gssLib = NULL;
+
+// static storage for all method pointers
+static gss_shim_t s_gss_shim;
+
+// reference to the shim storage.
+static gss_shim_t* volatile s_gss_shim_ptr = NULL;
```
Comment (janvorli) [inline src/libraries/Native/Unix/System.Net.Security.Native/pal_gssapi.c:104]: A nit - now that the shim is guaranteed to be initialized before the first use, we could just use a static instance of the gss_shim_t or just make the pointers standalone global variables without a wrapper struct like we do in the ICU and OpenSSL shims.

Comment (VSadov) [reply src/libraries/Native/Unix/System.Net.Security.Native/pal_gssapi.c:104]: Right, the indirection no longer serves the purpose of intercepting and initializing. I thought it was still useful for ordering the reads - to make sure individual proxies are not loaded before checking that the whole thing is initialized (on ARM64 and like that), but I guess I can just use an acquiring read.

Comment (VSadov) [reply src/libraries/Native/Unix/System.Net.Security.Native/pal_gssapi.c:104]: Actually I am not sure we need a flag - just let it initialize more than once. We should not have too many classes running this initializer anyways. 
As I see SSL does not try optimizing redundant initializations either, and it is a lot heavier there.

**Review by janvorli — APPROVED** [review]
LGTM, thank you!

**Comment by VSadov** [comment]
Thanks!!
